ipn/ipnlocal/local.go

```diff
@@ -3558,6 +3562,47 @@ func (b *LocalBackend) WaitingFiles() ([]apitype.WaitingFile, error) {
 	return apiSrv.WaitingFiles()
 }
 
+// mapSetHandle is a minimal (but non-zero) value whose addresse serves as a map
+// key for sets of non-comparable values that can't be map keys themselves.
+type mapSetHandle byte
+
+func (b *LocalBackend) setFileWaiter(handle *mapSetHandle, wakeWaiter context.CancelFunc) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	if wakeWaiter == nil {
+		delete(b.fileWaiters, handle)
+	} else {
+		mak.Set(&b.fileWaiters, handle, wakeWaiter)
+	}
+}
+
+// AwaitFile blocks until ctx is done or a file is available.
+// It reports whether it returned because a file is available.
+func (b *LocalBackend) AwaitFile(ctx context.Context) bool {
+	gotFile, gotFileCancel := context.WithCancel(context.Background())
+	defer gotFileCancel()
+
+	handle := new(mapSetHandle)
+	b.setFileWaiter(handle, gotFileCancel)
+	defer b.setFileWaiter(handle, nil)
+
+	// After we've registered our waiter, check if anything's available. This
+	// guarantees there's no window for a file to arrive between checking and
+	// registering our waiter and us blocking forever.
+	go func() {
+		if ff, _ := b.WaitingFiles(); len(ff) > 0 {
+			gotFileCancel()
+		}
```

Comment:
You could return the `ff` from `AwaitFile` here and simplify the `for` loop in `serveFiles`, I think?

Reply:
I didn't really want to reason about the races. Getting the list is cheap. I just want some cheap sleep+wake-up call.

Would you prefer it returned nothing instead? Then caller can check its own ctx.Err() that it passed in.

Reply:
Oh I think I know what you mean now. Let me try.

Reply:
PTAL